File: Apps4Bro.iOS/Apps4BroSDK.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Runtime.Remoting.Contexts;
using Google.UserMessagingPlatform;
using UIKit;

#if __ANDROID__
using Android.Gms.Ads.Identifier;
using Android.Gms.Common;

#else
#endif

namespace Apps4Bro
{
    public enum Apps4BroSettings
    {
        None = 0,
        ShowBannerOnTop = 1,
        UseBannerSuperview = 2
    }

    public static class Apps4BroSDK
    {
        private static readonly string s_emptyId = "00000000-0000-0000-0000-000000000000";
        public static readonly int Version = 130;

        private static string s_advertisingId = s_emptyId;
        private static bool s_inited = false;

        private static bool s_useNonPersonalizedAds = false;

        private static string s_platform = "unknown";

        internal readonly static string ReportUrl = "https://app4bro.runserver.net/app4bro/event.php?id={0}&app={1}&event={2}&param={3}&time={4}&eventid={5}";

        // /route/<App4Bro-ID>/ — path-embedded routing. Query params:
        //   id    = App4Bro app ID (same value as the path segment). Routing key.
        //   app   = analytics name (bundle identifier). Cosmetic — server uses it
        //           only in GA labels.
        //   lang/sdk/os = metadata.
        //   did   = device advertising ID. Trailing; may be empty when ATT denied
        //           / SDK not inited.
        internal readonly static string AdManagerUrl = "https://app4bro.runserver.net/route/{0}/?id={0}&app={1}&lang={2}&sdk={3}&os={4}&did={5}";
        internal readonly static string AdManagerUrlShort = "https://app4bro.runserver.net/route/{0}/";

        // HouseAdUrl args:
        //   {0}  id         — house zone ID (per-zone server config; NOT the App4Bro app ID)
        //   {1}  app        — analytics name (bundle identifier — same as AdManagerUrl `app=`)
        //   {2}  brand      — device brand ("Apple" on iOS)
        //   {3}  model      — device model identifier (e.g. "iPhone15,2")
        //   {4}  operator   — carrier (often empty on iOS 16+)
        //   {5}  width
        //   {6}  height
        //   {7}  lang       — ISO 639-1 language code
        //   {8}  sdk        — Apps4BroSDK.Version
        //   {9}  os         — platform name
        //   {10} osver      — OS version (e.g. "17.5")
        //   {11} appver     — host app version (CFBundleShortVersionString)
        //   {12} did        — device advertising ID (trailing; may be empty)
        internal readonly static string HouseAdUrl = "https://app4bro.runserver.net/app4bro/house.php?id={0}&app={1}&brand={2}&model={3}&operator={4}&width={5}&height={6}&lang={7}&sdk={8}&os={9}&osver={10}&appver={11}&did={12}";
        internal readonly static int HouseAdTimeout = 10;

        internal readonly static IDictionary<Apps4BroSettings, object> Settings = new Dictionary<Apps4BroSettings, object>();

        public static string Platform
        {
            get { return s_platform; }
        }

        public static bool UseNonPersonalizedAds
        {
            get { return s_useNonPersonalizedAds; }
            set { s_useNonPersonalizedAds = value; }
        }

        public static void InitCMP(UIViewController viewController)
        {
            var requestParameters = new RequestParameters();
            requestParameters.TagForUnderAgeOfConsent = false;

#if DEBUG
            requestParameters.DebugSettings = new DebugSettings
            {
                Geography = DebugGeography.Eea
            };
#endif

            ConsentStatus consentStatus = ConsentStatus.NotRequired;

            ConsentInformation.SharedInstance.RequestConsentInfoUpdate(requestParameters, (error) =>
            {
                if (error == null)
                {
                    consentStatus = ConsentInformation.SharedInstance.ConsentStatus;
                    // Check user location
                    if (consentStatus == ConsentStatus.Unknown
                    || consentStatus == ConsentStatus.Required) // IsRequestLocationInEeaOrUnknown
                    {
                        // Load consent message
                        ConsentForm.Load((consentForm, loadError) =>
                        {
                            if (loadError == null)
                            {
                                // Show consent message
                                consentForm.Present(viewController);
                            }
                        });
                    }
                }
            });
        }


        public static string AdvertisingId
        {
            get
            {

                if (!s_inited)
                {
#if __IOS__
                    Init(null); // we can init iOS sdk without context. Not possible for Android
#else
                    throw new ApplicationException("Apps4Bro SDK is not inited!");
#endif

                }

                if (string.IsNullOrEmpty(s_advertisingId) || s_advertisingId.Equals(s_emptyId))
                    return string.Empty; // conserve the traffic 

                return s_advertisingId;
            }
        }

        public static void SetSettings(Apps4BroSettings setting, object value)
        {
            Settings[setting] = value;
        }

        public static void Init(object context)
        {
            s_inited = true;
            try
            {
#if __IOS__

                s_platform = "ios";
                if (AdSupport.ASIdentifierManager.SharedManager.IsAdvertisingTrackingEnabled)
                    s_advertisingId = WebUtility.UrlEncode(AdSupport.ASIdentifierManager.SharedManager.AdvertisingIdentifier.AsString());

#elif __ANDROID__
                
                s_platform = Adroid.OS.Build.MANUFACTURER == "Amazon" ? "amazon" : "android";
                
                System.Threading.Thread thread= new System.Threading.Thread(delegate()
                    {
                        try
                        {
                            AdvertisingIdClient.Info adInfo = AdvertisingIdClient.GetAdvertisingIdInfo((Android.Content.Context)context);
                            s_advertisingId = WebUtility.UrlEncode(adInfo.Id);
                        }
                        catch(Exception ex)
                        {
                            Console.Error.WriteLine("Failed to get Advertising ID: " + ex);
                        }
                    }
                );
                thread.Start();
#elif NETFX_CORE
                s_platform = "win";
#endif
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("Failed to get Advertising ID: " + ex);
            }
        }
    }
}

File: Apps4Bro.Swift/Sources/Apps4Bro/Apps4BroSDK.swift
import Foundation
import UIKit
import AdSupport
import AppTrackingTransparency
import UserMessagingPlatform // module ships under this name from the GoogleUserMessagingPlatform SPM package

// Port of Apps4BroSDK.cs. C# was a static class — Swift keeps the same API as
// type-level statics on `Apps4BroSDK`.
public enum Apps4BroSettings: Int {
    case none = 0
    case showBannerOnTop = 1
    case useBannerSuperview = 2
}

public enum Apps4BroSDK {
    private static let emptyId = "00000000-0000-0000-0000-000000000000"

    /// Bumped from C# Version=130. Keep in sync if the server expects parity.
    public static let version: Int = 130

    private static var advertisingIdValue: String = emptyId
    private static var initedFlag: Bool = false
    private static var platformValue: String = "unknown"

    public static var useNonPersonalizedAds: Bool = false

    public static var platform: String { platformValue }

    /// Server endpoints — mirrors C# `internal readonly static string` constants.
    /// `settings` and `useNonPersonalizedAds` are read by the network implementations.
    static let reportUrl = "https://app4bro.runserver.net/app4bro/event.php?id=%@&app=%@&event=%@&param=%@&time=%d&eventid=%d"
    // /route/<App4Bro-ID>/ — path-embedded routing (dodges ad-blockers that catch `/ad`).
    // Query params:
    //   id    = App4Bro app ID (same value as path segment). Routing key.
    //   app   = analytics name (bundle identifier). Cosmetic — server uses it only
    //           in GA labels.
    //   lang/sdk/os = metadata.
    //   did   = device advertising ID. Trailing; may be empty when ATT denied.
    //
    // Args order across SDKs (positional): appId, analyticsName, lang, sdk, os, did.
    static let adManagerUrl = "https://app4bro.runserver.net/route/%1$@/?id=%1$@&app=%2$@&lang=%3$@&sdk=%4$d&os=%5$@&did=%6$@"
    static let adManagerUrlShort = "https://app4bro.runserver.net/route/%@/"
    // HouseAdUrl args (in positional order):
    //   id         — house zone ID (per-zone server config; NOT the App4Bro app ID)
    //   app        — analytics name (bundle identifier — same convention as AdManagerUrl)
    //   brand      — device brand ("Apple" on iOS)
    //   model      — device model identifier (e.g. "iPhone15,2")
    //   operator   — carrier (often empty on iOS 16+)
    //   width, height
    //   lang       — ISO 639-1 language code
    //   sdk        — Apps4BroSDK.Version
    //   os         — platform name
    //   osver      — OS version (e.g. "17.5")
    //   appver     — host app version (CFBundleShortVersionString)
    //   did        — device advertising ID (trailing; may be empty)
    //   noclose    — true → hide close button (banner), false → show it (interstitial)
    static let houseAdUrl = "https://app4bro.runserver.net/app4bro/house.php?id=%@&app=%@&brand=%@&model=%@&operator=%@&width=%d&height=%d&lang=%@&sdk=%d&os=%@&osver=%@&appver=%@&did=%@&noclose=%@"
    static let houseAdTimeout: TimeInterval = 10

    private static let settingsQueue = DispatchQueue(label: "Apps4Bro.settings")
    private static var settingsStorage: [Apps4BroSettings: Any] = [:]
    static var settings: [Apps4BroSettings: Any] {
        settingsQueue.sync { settingsStorage }
    }

    public static func setSetting(_ setting: Apps4BroSettings, value: Any) {
        settingsQueue.sync { settingsStorage[setting] = value }
    }

    /// C# `AdvertisingId` getter — auto-inits on iOS if the host never called `init`.
    public static var advertisingId: String {
        if !initedFlag {
            initialize(context: nil)
        }
        if advertisingIdValue.isEmpty || advertisingIdValue == emptyId {
            return "" // C# behavior: conserve the traffic
        }
        return advertisingIdValue
    }

    /// Port of `Init(object context)`. Matches C# semantics: idempotent flag, reads
    /// the IDFA via ASIdentifierManager. C# checked `IsAdvertisingTrackingEnabled`
    /// — deprecated since iOS 14; the modern equivalent is App Tracking Transparency.
    /// Hosts that want a real IDFA must call `requestTrackingAuthorization` first.
    public static func initialize(context: Any?) {
        initedFlag = true
        platformValue = "ios"

        let manager = ASIdentifierManager.shared()
        let authorized: Bool
        if #available(iOS 14, *) {
            authorized = ATTrackingManager.trackingAuthorizationStatus == .authorized
        } else {
            // iOS 13: fall back to the deprecated isAdvertisingTrackingEnabled.
            authorized = manager.isAdvertisingTrackingEnabled
        }
        if authorized {
            let raw = manager.advertisingIdentifier.uuidString
            advertisingIdValue = raw.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? raw
        }
    }

    /// Port of `InitCMP` — drives the Google UserMessagingPlatform consent flow.
    public static func initCMP(viewController: UIViewController) {
        let params = UMPRequestParameters()
        params.tagForUnderAgeOfConsent = false
        #if DEBUG
        let debug = UMPDebugSettings()
        debug.geography = .EEA
        params.debugSettings = debug
        #endif

        UMPConsentInformation.sharedInstance.requestConsentInfoUpdate(with: params) { error in
            if let error = error {
                Log.w("Apps4BroSDK", "UMP consent info update failed: \(error)")
                return
            }
            let status = UMPConsentInformation.sharedInstance.consentStatus
            guard status == .unknown || status == .required else { return }
            UMPConsentForm.load { form, loadError in
                if let loadError = loadError {
                    Log.w("Apps4BroSDK", "UMP consent form load failed: \(loadError)")
                    return
                }
                guard let form = form else { return }
                form.present(from: viewController) { _ in }
            }
        }
    }
}
